src/client/dbus/thread_api_dbus.cpp

```diff
@@ -0,0 +1,510 @@
+        {
+            return p.second;
+        }
+    }
+
+    assert(false);
+}
+
+namespace otbr {
+namespace client {
+
+bool IsThreadActive(otDeviceRole aRole)
+{
+    bool isActive = false;
+
+    switch (aRole)
+    {
+    case OT_DEVICE_ROLE_DISABLED:
+    case OT_DEVICE_ROLE_DETACHED:
+        isActive = false;
+        break;
+    case OT_DEVICE_ROLE_CHILD:
+    case OT_DEVICE_ROLE_ROUTER:
+    case OT_DEVICE_ROLE_LEADER:
+        isActive = true;
+        break;
+    }
+
+    return isActive;
+}
+
+ThreadApiDBus::ThreadApiDBus(DBusConnection *aConnection)
+    : mInterfaceName("wpan0")
+    , mConnection(aConnection)
+{
+    SubscribeDeviceRoleSignal();
+}
+
+ThreadApiDBus::ThreadApiDBus(DBusConnection *aConnection, const std::string &aInterfaceName)
+    : mInterfaceName(aInterfaceName)
+    , mConnection(aConnection)
+{
+    SubscribeDeviceRoleSignal();
+}
+
+otError ThreadApiDBus::SubscribeDeviceRoleSignal(void)
+{
+    std::string matchRule = "type='signal',interface='" DBUS_INTERFACE_PROPERTIES "'";
+    DBusError   err;
+    otError     ret = OT_ERROR_NONE;
+
+    dbus_error_init(&err);
+    dbus_bus_add_match(mConnection, matchRule.c_str(), &err);
+
+    VerifyOrExit(!dbus_error_is_set(&err), ret = OT_ERROR_FAILED);
+
+    dbus_connection_add_filter(mConnection, sDBusMessageFilter, this, nullptr);
+exit:
+    printf("SubscribeDeviceRoleSignal err %d\n", ret);
+    return ret;
+}
+
+DBusHandlerResult ThreadApiDBus::sDBusMessageFilter(DBusConnection *aConnection, DBusMessage *aMessage, void *aData)
+{
+    ThreadApiDBus *api = static_cast<ThreadApiDBus *>(aData);
+
+    return api->DBusMessageFilter(aConnection, aMessage);
+}
+
+DBusHandlerResult ThreadApiDBus::DBusMessageFilter(DBusConnection *aConnection, DBusMessage *aMessage)
+{
+    (void)aConnection;
+
+    if (dbus_message_is_signal(aMessage, DBUS_INTERFACE_PROPERTIES, DBUS_PROPERTIES_CHANGED_SIGNAL))
+    {
+        DBusMessageIter iter, subIter, dictEntryIter, valIter;
+        std::string     interfaceName, propertyName, val;
+
+        VerifyOrExit(dbus_message_iter_init(aMessage, &iter));
+        SuccessOrExit(DBusMessageExtract(&iter, interfaceName));
+        VerifyOrExit(interfaceName == OTBR_DBUS_PREFIX);
+
+        VerifyOrExit(dbus_message_iter_get_arg_type(&iter) == DBUS_TYPE_ARRAY);
+        dbus_message_iter_recurse(&iter, &subIter);
+        VerifyOrExit(dbus_message_iter_get_arg_type(&subIter) == DBUS_TYPE_DICT_ENTRY);
+        dbus_message_iter_recurse(&subIter, &dictEntryIter);
+        SuccessOrExit(DBusMessageExtract(&dictEntryIter, propertyName));
+        VerifyOrExit(dbus_message_iter_get_arg_type(&dictEntryIter) == DBUS_TYPE_VARIANT);
+        dbus_message_iter_recurse(&dictEntryIter, &valIter);
+        SuccessOrExit(DBusMessageExtract(&valIter, val));
+
+        VerifyOrExit(propertyName == OTBR_DBUS_DEVICE_ROLE_PROPERTY);
+
+        for (auto &&f : mDeviceRoleHandlers)
+        {
+            f(NameToDeviceRole(val));
```

Comment:
Why rvalue reference here?

Reply:
It's a universal reference other than a rvalue reference. A const auto& maybe a better bit here.

Reply:
I will suggest `const auto&`.